model/flow/order/identity.go

```diff
@@ -7,41 +7,29 @@ import (
 )
 
 // Canonical represents the canonical ordering for identity lists.
-func Canonical(identity1 *flow.Identity, identity2 *flow.Identity) bool {
+func Canonical(identity1 *flow.Identity, identity2 *flow.Identity) int {
 	return IdentifierCanonical(identity1.NodeID, identity2.NodeID)
 }
 
 // ByReferenceOrder return a function for sorting identities based on the order
 // of the given nodeIDs
-func ByReferenceOrder(nodeIDs []flow.Identifier) func(*flow.Identity, *flow.Identity) bool {
-	indices := make(map[flow.Identifier]uint)
+func ByReferenceOrder(nodeIDs []flow.Identifier) func(*flow.Identity, *flow.Identity) int {
+	indices := make(map[flow.Identifier]int)
 	for index, nodeID := range nodeIDs {
 		_, ok := indices[nodeID]
 		if ok {
 			panic("should never order by reference order with duplicate node IDs")
 		}
-		indices[nodeID] = uint(index)
+		indices[nodeID] = index
 	}
-	return func(identity1 *flow.Identity, identity2 *flow.Identity) bool {
-		return indices[identity1.NodeID] < indices[identity2.NodeID]
+	return func(identity1 *flow.Identity, identity2 *flow.Identity) int {
+		return indices[identity1.NodeID] - indices[identity2.NodeID]
 	}
 }
 
 // IdentityListCanonical takes a list of identities and
-// check if it's ordered in canonical order.
+// check if it's strictly ordered in canonical order.
+// Strict ordering means that equality in ordering is not allowed.
 func IdentityListCanonical(identities flow.IdentityList) bool {
```

Comment:
following up on [my earlier comment](), I  would suggest to:
* rename this function to `IsIdentityListCanonical`
* make the relation between canonical ordering and the return values unambiguously clear in the goDoc.

```suggestion
// IdentityListCanonical returns true if and only if the given IdentityList is in
// canonical order.By convention, two Identities (i1, i2) are in canonical order
// if i1's NodeID is lexicographically _strictly_ smaller than i2's NodeID.
// The strictness is important here, meaning that we do not allow duplicates. In other words,
// (i,i) does not satisfy canonical ordering.
func IsIdentityListCanonical(identities flow.IdentityList) bool {
```

Reply:
I added your comment with a little modification. There is a distinction in the definition of "ordered" and "sorted" that I would like to make:

- an order is a [binary relation](https://en.wikipedia.org/wiki/Binary_relation), it means it involves 2 elements only. The relation becomes an "order" once some properties are satisfied. We can therefore talk about 2 elements being in a certain order (in our case the defined canonical order). 
- a sort involves a list of elements, which are re-listed with regards to some order defined between 2 elements. The [sort we use](https://pkg.go.dev/golang.org/x/exp/slices#SortFunc) requires that the binary order relation is a strict weak ordering, which implies (among other things) that the order is irreflexive (an element `x` isn't in order with itself). This is the case of our canonical ordering (if we note our order `<`, for all identifiers the statement `x < x` is false, or in other words `x` and `x` aren't in a canonical order), so it's fine to use it for sorting. From the perspective of the sorting function, it is not required that the elements to be sorted are all distinct. For instance the sort result of [4,1,3,1] with regards to the integer strict inequality `<` is [1,1,3,4].  This final list is well-sorted with regards to `<`, even though `<` is irreflexive, just like our canonical order.  Sorting only makes sure that there are no successive elements `list[i]` and `list[i+1]` such that `list[i+1] < list[i]`. So what I want to say here is:
  - I would describe a list as "sorted with regards to a canonical order" and not "in canonical order". The latter expression is reserved to two elements.
  - in our case, "sorted with regards to a canonical order" isn't enough to describe what we want the function to check. We're also adding the "strict sorting", which means that we want the function to make sure that no 2 elements from the list are equal, in addition to the list being sorted.

Reply:
[more clarification and test minor update](https://github.com/onflow/flow-go/pull/5155/commits/0f27baaf10df7d9eab7b42f1d82dc4d0ee70e9c0)